src/Makefile

```diff
@@ -1,52 +1,80 @@
+ifeq ($(shell which llvm-config-12),)
+  ifeq ($(shell which llvm-config-9),)
+    ifeq ($(shell which llvm-config),)
+      $(error Unable to locate the LLVM toolchain. \
+        Check your $$PATH or reinstall. Supported LLVM versions are 9 and 12)
+    else
+      LLVM_CONFIG = llvm-config
+      CLANG_FORMAT = clang-format
+    endif
+  else
+    LLVM_CONFIG = llvm-config-9
+    CLANG_FORMAT = clang-format-9
+  endif
+else
+  LLVM_CONFIG = llvm-config-12
+  CLANG_FORMAT = clang-format-12
+endif
+
+LLVM_VERSION := $(shell $(LLVM_CONFIG) --version | cut -d '.' -f 1)
+ifeq ($(LLVM_VERSION), 12)
+else ifeq ($(LLVM_VERSION), 9)
+  LLVM = -DLLVM9
+else
+  $(error LLVM $(LLVM_VERSION) is not supported. Supported LLVM versions are 9 and 12)
+endif
+
+$(info Found LLVM $(LLVM_VERSION) toolchain)
+
 ifeq ($(shell uname), Darwin)
-LLVM_CONFIG=llvm-config
-LOADABLE_MODULE_OPTIONS=-bundle -undefined dynamic_lookup
+  LOADABLE_MODULE_OPTIONS = -bundle -undefined dynamic_lookup
 else
-LLVM_CONFIG=llvm-config-12
-LOADABLE_MODULE_OPTIONS=-shared -O1
+  LOADABLE_MODULE_OPTIONS = -shared -O1
 endif
 
 ifndef VERBOSE
-	QUIET:=@
+  QUIET:=@
 endif
 
-SRC_DIR?=$(PWD)
-LIB_DIR=$(SRC_DIR)/../lib
+SRC_DIR := $(shell pwd)
+LIB_DIR = $(SRC_DIR)/../lib
 
-COMMON_FLAGS=-g -Wall -Wextra -Wno-unused-parameter -Wno-implicit-fallthrough -fPIC
-LDFLAGS+=$(shell $(LLVM_CONFIG) --ldflags)
-CXXFLAGS+=$(COMMON_FLAGS) $(shell $(LLVM_CONFIG) --cxxflags)
-CPPFLAGS+=$(shell $(LLVM_CONFIG) --cppflags) -I$(SRC_DIR)
+COMMON_FLAGS = -g -Wall -Wextra -Wno-unused-parameter -Wno-implicit-fallthrough -fPIC
+LDFLAGS += $(shell $(LLVM_CONFIG) --ldflags)
+CXXFLAGS += $(COMMON_FLAGS) $(shell $(LLVM_CONFIG) --cxxflags)
+CPPFLAGS += $(shell $(LLVM_CONFIG) --cppflags) -I$(SRC_DIR)
 
-CI_PASS_OBJECTS=CompilerInterrupt.o
-CI_PASS=CompilerInterrupt.so
+CI_PASS_OBJECT = CompilerInterrupt.o
+CI_PASS = CompilerInterrupt.so
 
-CI_LIB_OBJECTS=ci_lib.o
-CI_LIB_DYN=libci.so
-CI_LIB_STATIC=libci.a
+CI_LIB_OBJECT = ci_lib.o
+CI_LIB_DYN = libci.so
+CI_LIB_STATIC = libci.a
 
-default: $(LIB_DIR)/$(CI_PASS) $(LIB_DIR)/$(CI_LIB_DYN) $(LIB_DIR)/$(CI_LIB_STATIC)
+all: $(LIB_DIR)/$(CI_PASS) $(LIB_DIR)/$(CI_LIB_DYN) $(LIB_DIR)/$(CI_LIB_STATIC)
 
 %.o: $(SRC_DIR)/%.cpp
-	@echo "Compiling $*.cpp"
-	clang-format -i $<
-	$(QUIET)$(CXX) -c $(CPPFLAGS) $(CXXFLAGS) $<
+	$(info Compiling $*.cpp)
+	$(QUIET)$(CLANG_FORMAT) -style=llvm -i $(wildcard *.cpp) $(wildcard *.hpp)
```

Comment:
Not using any .hpp files here. Seems redundant.

Reply:
Updated in b45282c.